smarts/core/lanepoints.py

```diff
@@ -0,0 +1,564 @@
+# Copyright (C) 2020. Huawei Technologies Co., Ltd. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+from __future__ import (  # to allow for typing to refer to class being defined (LinkedLanePoint)
+    annotations,
+)
+
+import math
+import queue
+from collections import defaultdict
+from dataclasses import dataclass
+from functools import lru_cache
+from typing import List, NamedTuple
+
+import numpy as np
+
+import warnings
+
+with warnings.catch_warnings():
+    from smarts.core.utils.file import suppress_pkg_resources
+
+    warnings.filterwarnings("ignore", "numpy.ufunc size changed")
+    with suppress_pkg_resources():
+        # We force sklearn to use a different importer since sklearn's use of pkg_resources is
+        # aggressive
+        from sklearn.neighbors import KDTree
+
+from smarts.core.coordinates import Heading, Pose
+from smarts.core.utils.math import (
+    lerp,
+    radians_to_vec,
+    signed_dist_to_line,
+    squared_dist,
+    vec_2d,
+    vec_to_radians,
+)
+
+
+@dataclass(frozen=True)
+class LanePoint:
+    """a.k.a. MapPoint:  static, based only on a map, always in a Lane"""
+
+    pos: np.ndarray  # Point positioned on center of lane
+    heading: Heading  # Heading angle of lane at this point (radians)
+    lane_width: float  # Width of lane at this point (meters)
+    speed_limit: float  # Lane speed in m/s
+    lane_id: str  # ID of lane under lanepoint
+    lane_index: int  # Index of the lane this lanepoint is over. 0 is the outer(right) most lane.
+
+    def dist_to(self, p) -> float:
```

Comment:
Type of `p` ?

Reply:
Other than knowing it's potentially some "higher order" vector, I honestly have no idea!  (I didn't write this code originally, just moved it from Waypoints.)  I tried tracing a bit, both up and down the callstack, but the type wasn't obvious, so I left it as it was.